### ISSUE TITLE
Linux Flatpak

### ISSUE THREAD

**Issue by joltman80** [zathras/jrpn#128]
**Is your feature request related to a problem? Please describe.**
This is not related to a bug.

**Describe the solution you'd like**
I don't run Ubuntu, and I'd like to avoid installing snapd.  Would it be possible to bundle JRPN as an official flatpak as well?

**Describe any alternatives you've considered**
There seem to be several "wrapped application" solutions out there.  Snapd is mostly bound to Ubuntu, while Flatpak appears to be more universally available across distributions.

**Additional context**
I know it's extra work, but it would make JRPN more widely available.  Thank you!

**Comment by zathras** [comment]
Well, I looked up what "flatpak" is.  It seems its both a distribution format, and also it sometimes means the main app store that distributes flatpaks.  Which do you mean - a flatpak binary, say on github, or a listing on the flatpak app store?  Also, where is that app store?  Apparently there are more than one, but there's a main one that most people use, if I remember correctly.

I'm not opposed to having Yet Another Distribution Channel, though I would put a high premium on something I can more easily automate.  Ubuntu's is pretty painless.

**Comment by zathras** [comment]
You're welcome!  It's a fun little project.

If I were to try to create something on flathub.org, would you be able to download and test it?  I have a couple of old Linux computers floating around, but installing from a different app store ecosystem would add steps, and there's always the danger that it would accidentally work because I have the Flutter development platform installed.  So if ;you could do a smoke test on a beta listing, I might be able to take a look at this, maybe in the first quarter of next year.  These CI/CD workflows always take some time to set up.

Note to self:  Here are some other resources.
https://github.com/flatpak/flatpak-github-actions https://github.com/Merrit/flutter_flatpak_example

**Comment by zathras** [comment]
Oh, out of curiosity, does installing it manually from https://github.com/zathras/jrpn/releases/download/2.1.16/jrpn15_linux.zip work?  It should just be a matter of unpacking that somewhere, then making an alias or whatever to get at the shell script that launches it.  Of course that won't auto-update when I do a new release, and it doesn't run it in a container, so I guess it doesn't have the same level of security guarantee.

**Comment by joltman80** [comment]
I would be happy to be a beta tester.  I've not done something like that before, but I'll do what you need.  I've just downloaded the ZIP file above, unzipped into ~/Applications/jrpn15 and from terminal ran ./jrpn15 and the app started correctly.  Performing a simple calculation yielded the expected result.

As an FYI, I'm running Arch Linux, kernel 6.12.3, with Gnome 47 on Wayland.  When going to HELP > SEE CALCULATOR INTERNALS, there's a message "On the Linux snap, closing this window exits calculator."  When I exit that INTERNALS window, the program also closes.  Let me know what else you need.  Thanks!

**Comment by zathras** [comment]
Great, thanks!

> As an FYI, I'm running Arch Linux, kernel 6.12.3, with Gnome 47 on Wayland. When going to HELP > SEE CALCULATOR INTERNALS, there's a message "On the Linux snap, closing this window exits calculator." When I exit that INTERNALS window, the program also closes.

Ah, so it's not specific to the snap.  I guess I'll have to upgrade the message to say it's on Linux generally.

The issue is that Google's Flutter doesn't really support multiple windows.  It's been on their "to do" list for a couple of years, but Google has made cutbacks to the team, and desktop isn't a priority.  I did find a third-party library for multiple windows that sort of works, but it's not very good quality.  On Linux, the snap terminates, but when I run it in development mode, it doesn't.  So, my choices on Linux are to disable the feature completely, and make it take over the window like it does on Android, or give that warning message.  I went with the warning.

Also, since I don't really trust that library and it has native code, I forked my own copy and audited it, to make sure there was no malware.

Anyway, when I get a chance to dive into Github actions and flatpaks, I'll be in touch.

**Comment by joltman80** [comment]
Sounds good!  And I only wanted to report the fact that the message says "snap" in it, so that if you wanted to modify the message, you could.  I'll be around for any messages.  Thanks!

**Comment by zathras** [comment]
Created https://github.com/flathub/flathub/pull/10348

**Comment by joltman80** [comment]
I looked at your PR into Flathub.  While you checked the box, it looks like you forgot to answer the disclosure of AI generated material question.

> I have disclosed any AI-generated material included in the application or its Flathub packaging, as required by the [Generative AI policy](https://docs.flathub.org/docs/for-app-authors/requirements#generative-ai-policy). Affected parts and approximate extent: < Please describe here, or write None >

I'm guessing you should enter NONE.

**Comment by teamcons** [comment]
> From e-mail:
> 
> > Now to the issue.  I am running Linux Mint 22.2 Xfce on my laptop.  I just installed JRPN15 via snap.  There doesn't seem to be any way to launch the app graphically.  The only way I have found is to do an app search for jrpn15 and then click on the resulting "Run jrpn15".  Am I missing something?
> 
> I don't actually know, but probably not? I don't use Linux all that regularly, but it seems that people are unhappy with the Ubuntu snap store. On Ubuntu linux, it _does_ give you an icon, I think, though it's been several years since I did the installation on that old computer. It seems that "flatpak" is the favored installation option on Mint, and possibly others? I did just install Mint on a computer I'm loaning to a friend, so maybe I'll try to figure out what it takes to package the app this way.

probably because snap does not enforce the xdg standards like flatpak does

On Linux, to be considered a "app with GUI", the app should have a .desktop file, which describes what name it has, what command to start the app, what icon to display when it is started, and what right-click options it has

in the proper folder, it allows for example the application view in ubuntu to display it properly, and not skip it like the ten billions of commandline-only utilities linuxes have

**Comment by zathras** [comment]
> I looked at your PR into Flathub. While you checked the box, it looks like you forgot to answer the disclosure of AI generated material question.
> 
> > I have disclosed any AI-generated material included in the application or its Flathub packaging, as required by the [Generative AI policy](https://docs.flathub.org/docs/for-app-authors/requirements#generative-ai-policy). Affected parts and approximate extent: < Please describe here, or write None >
> 
> I'm guessing you should enter NONE.

Thanks, yes.  This app predates AI slop by a fair bit.

**Comment by zathras** [comment]
> probably because snap does not enforce the xdg standards like flatpak does
> 
> On Linux, to be considered a "app with GUI", the app should have a .desktop file, which describes what name it has, what command to start the app, what icon to display when it is started, and what right-click options it has
> 
> in the proper folder

Right, thanks, I'm familiar with .desktop files.  What wasn't clear however many years ago I set up the snap was that snap didn't handle this automatically, or where they wanted me to put the .desktop file.

**Comment by zathras** [comment]
https://github.com/o-murphy/flutpak/issues/57
